### ISSUE TITLE
(t) Failing unit tests

### ISSUE THREAD

**Issue by phillxnet** [rockstor/rockstor-core#2371]
(t) = testing branch. Prior to our recent large Django version update we had all unite tests passing.
We now have:
```
Ran 208 tests in 19.686s

FAILED (failures=4, errors=10)
```

Noting the following failure/Error overview captured via:

> ./bin/test > test-results.txt 2>&1

>.E..............................F....................................E...E...............E.......E.E..E...................E....E............E...FFF......................................................................

## Error results:

- [x] ERROR: test_get (storageadmin.tests.test_appliances.AppliancesTests) -
- TypeError: assertEqual() got an unexpected keyword argument 'follow'
- [ ] ERROR: test_post_requests_2 (storageadmin.tests.test_pool_balance.PoolBalanceTests)
- IntegrityError: insert or update on table "storageadmin_poolbalance" violates foreign key constraint "storageadmin_poolbal_pool_id_f6813dfc_fk_storagead"
DETAIL:  Key (pool_id)=(2) is not present in table "storageadmin_pool".
- [ ] ERROR: test_post_requests_2 (storageadmin.tests.test_pool_scrub.PoolScrubTests)
- IntegrityError: insert or update on table "storageadmin_poolscrub" violates foreign key constraint "storageadmin_poolscrub_pool_id_9910cdd4_fk_storageadmin_pool_id"
DETAIL:  Key (pool_id)=(2) is not present in table "storageadmin_pool".
- [ ] ERROR: test_create_samba_share (storageadmin.tests.test_samba.SambaTests)
- IntegrityError: insert or update on table "storageadmin_sambashare" violates foreign key constraint "storageadmin_sambash_share_id_e755b23e_fk_storagead"
DETAIL:  Key (share_id)=(23) is not present in table "storageadmin_share".
- [ ] ERROR: test_post_requests_2 (storageadmin.tests.test_samba.SambaTests)
- IntegrityError: insert or update on table "storageadmin_sambashare" violates foreign key constraint "storageadmin_sambash_share_id_e755b23e_fk_storagead"
DETAIL:  Key (share_id)=(24) is not present in table "storageadmin_share".
- [ ] ERROR: test_post_requests_no_admin (storageadmin.tests.test_samba.SambaTests)
- IntegrityError: insert or update on table "storageadmin_sambashare" violates foreign key constraint "storageadmin_sambash_share_id_e755b23e_fk_storagead"
DETAIL:  Key (share_id)=(23) is not present in table "storageadmin_share".
- [ ] ERROR: test_put_requests_2 (storageadmin.tests.test_samba.SambaTests)
- IntegrityError: insert or update on table "storageadmin_sambashare" violates foreign key constraint "storageadmin_sambash_share_id_e755b23e_fk_storagead"
DETAIL:  Key (share_id)=(23) is not present in table "storageadmin_share".
- [ ] ERROR: test_clone_command (storageadmin.tests.test_snapshot.SnapshotTests)
- IntegrityError: insert or update on table "storageadmin_share" violates foreign key constraint "storageadmin_share_pool_id_e63524ad_fk_storageadmin_pool_id"
DETAIL:  Key (pool_id)=(1) is not present in table "storageadmin_pool".
- [ ] ERROR: test_post_requests_2 (storageadmin.tests.test_snapshot.SnapshotTests)
- IntegrityError: insert or update on table "storageadmin_share" violates foreign key constraint "storageadmin_share_pool_id_e63524ad_fk_storageadmin_pool_id"
DETAIL:  Key (pool_id)=(1) is not present in table "storageadmin_pool".
- [ ] ERROR: setUpClass (smart_manager.tests.test_snmp.SNMPTests)
- IntegrityError: Problem installing fixture '/opt/rockstor-dev/src/rockstor/smart_manager/fixtures/services.json': Could not load oauth2_provider.Application(pk=1): null value in column "created" violates not-null constraint
DETAIL:  Failing row contains (1, y5s_6Ge1yP.xUacU_U9_;t.bdAosyX43a2Of.kdH, , confidential, client-credentials, wIsVYk0=14O?JylV.Y7p2lcOnIeWmC_lZEuKy@xloaVj9dIQol@5Rp.V7M4gx2PS..., cliapp, 1, f, null, null).

## Fail results:

- [ ] FAIL: test_post_requests_2 (storageadmin.tests.test_disk_smart.DiskSmartTests)
- AssertionError: ['Failed to insert expression "<MagicMock name=\'info().__getitem__().resolve_expression()\' id=\'140611611569744\'>" on storageadmin.SMARTIdentity.model_family. F() expressions can only be used to update, not to insert.'
- [ ] FAIL: test_post_invalid_type (smart_manager.tests.test_task_scheduler.TaskSchedulerTests)
- mass of html
- [ ] FAIL: test_post_name_exists (smart_manager.tests.test_task_scheduler.TaskSchedulerTests)
- mass of html
- [ ] FAIL: test_post_valid (smart_manager.tests.test_task_scheduler.TaskSchedulerTests)
- mass of html

**Comment by phillxnet** [comment]
Initial ERROR from incorrect mass insertion of follow=true.
We now have another failure instead of a test error:

- [ ] FAIL: test_get (storageadmin.tests.test_appliances.AppliancesTests)

`AssertionError: ['Appliance matching query does not exist.', 'Traceback (most recent call last):\n  File "/opt/rockstor-dev/src/rockstor/rest_framework_custom/generic_view.py", line 41, in _handle_exception\n    yield\n  File "/opt/rockstor-dev/src/rockstor/storageadmin/views/appliances.py", line 42, in get_queryset\n    self._update_hostname()\n  File "/opt/rockstor-dev/eggs/Django-1.11.29-py2.7.egg/django/utils/decorators.py", line 185, in inner\n    return func(*args, **kwargs)\n  File "/opt/rockstor-dev/src/rockstor/storageadmin/views/appliances.py", line 49, in _update_hostname\n    a = Appliance.objects.get(current_appliance=True)\n  File "/opt/rockstor-dev/eggs/Django-1.11.29-py2.7.egg/django/db/models/manager.py", line 85, in manager_method\n    return getattr(self.get_queryset(), name)(*args, **kwargs)\n  File "/opt/rockstor-dev/eggs/Django-1.11.29-py2.7.egg/django/db/models/query.py", line 380, in get\n    self.model._meta.object_name\nDoesNotExist: Appliance matching query does not exist.\n']`

**Comment by phillxnet** [comment]
@FroggyFlox & @Hooverdan96 the vast majority of our existing error test results are due to:

IntegrityError: insert or update on table "storageadmin_X-TABLE" violates foreign key constraint "storageadmin_*arbitrary-reference-here*"
DETAIL: Key (share_id)=(23) is not present in table "storageadmin_Y-TABLE".

Where the affected TABLES are:

| on table (X-TABLE) | is not present in table (Y-TABLE) |
| - | - |
| storageadmin_poolbalance | storageadmin_pool |
| storageadmin_poolscrub |  storageadmin_pool |
| storageadmin_sambashare |  storageadmin_share |
| storageadmin_share |  storageadmin_pool |

This looks to be a new constraint regarding Django 1.10 which concurs with our move from 1.8.16 -> 1.11.29
For reference see:
https://code.djangoproject.com/ticket/27315

Noting as this is as far as I've got with this and I'm a bit stumped.

**Comment by Hooverdan96** [comment]
Not to go down the wrong path entirely, but I did notice when I was doing my extreme upgrade that I was getting a lot of errors up front because of missing on_delete attributes for ForeignKey relationships. Not sure that applies to this upgraded version as well. In the end it seemed I had to remove the original migrations and create new ones from scratch ...
Looking at how the Foreign Key changes across versions, does this play a role?
Foreign Key 1.8: https://docs.djangoproject.com/en/1.8/ref/models/fields/#django.db.models.ForeignKey
Foreign Key 1.11: https://docs.djangoproject.com/en/1.11/ref/models/fields/#django.db.models.ForeignKey
Foreign Key 4.0: https://docs.djangoproject.com/en/4.0/ref/models/fields/#django.db.models.ForeignKey.db_constraint

Or am I missing that this is purely a test case issue?

For my own understanding I superimposed the db tables and class definitions from the corresponding models (tables based on the current 4.0.1 version) to get a better view of how things are connected.


<img width="1230" alt="Pool-Shares" src="https://user-images.githubusercontent.com/35113775/164916676-0af54ab6-1ae0-44ee-8d0a-227f3edc3214.png">

**Comment by phillxnet** [comment]
@Hooverdan96 Thanks for the pointers.
Re:
> I was getting a lot of errors up front because of missing on_delete attributes for ForeignKey relationships.
Yes I got these too and as you say they were up-front. From memory I thought I had fixed at the time (in #2365) by changing the model definitions. However I don't actually see this change there! So that's curious. I'll check again. Or it is actually a later Django thing that I'm remembering from Appman. 

> Or am I missing that this is purely a test case issue?

I believe it is yes. Most stuff is actually working but we have issues currently with smart-manager I think so wanted to address the test some more to have more insight on the remaining issues. And with many erring entirely they arn't that helpful in their relevant areas. 

I.e. most of our failing page are within the following:
https://github.com/rockstor/rockstor-core/blob/testing/src/rockstor/urls.py#L23-L28
So that should narrow down things down. But it's not quite speaking to me just yet. Hence the effort here to get the tests back up and running, or at least not ending in Error.

Thanks for adding the diagram, nice to take another look at them. PyCharm does a nice job of visualising these. I think it's in the Pro version only though.

I'm edging towards thinking we have what is now considered insufficient interconnects in our fixtures and mocked models. See that reference I posted where the report mirrored ours re tests working in 1.9 and not in 1.10. Where as previously these inter model links were ignored if they didn't exist. I.e. we are just failing increased sanity checks within our known old fixtures and model mocking. And given some of our fixtures contain canonical disk models such as "sda" when we moved years ago now to by-id type disk names we should probably just step up to recreating new fixtures. Bit of a pain but I think it's the most likely cause here. It may also be some kind of cache change from a little reading I did.

Re your own extreme efforts on the massive Django update you started (1.9 -> 4.04); how do you fancy doing a re-base on current testing channel and presenting as a pr. We could then more easily tease out from it some future changes that you may well have lined up for us. And this re-base may well also highlight some alternatives to my own minor updates thus far. I.e. in serializer.py I opted for the quick and dirty 'fields = "__all__"' for instance. As stated in our side channel chat I'd rather go in much smaller steps: because as we see here there are many subtle differences that come along with each Django update that we really have to catch as we go. And then once sorted we can tackle the next lot. But primarily our options for all things open up far wider once we are in Python 3 land. And testings current 1.11.29 Django, although old, is Python 3 compatible so we can in turn then jump up our Python before our next Django jump. Anyway I had meant to ask if you would consider doing the re-base and see how it goes. If nothing else it will help to see where we have to go in the more medium term. Although even then we are likely sticking to LTS Django only, just due to their life cycle being more appropriate to our current developer team count ;).

Thanks again for taking a look at this. I'm hoping to get back on this one soon as it would be super helpful to have these tests all back in play. Almost there now however.

**Comment by Hooverdan96** [comment]
@phillxnet, looks like you are narrowing it down. I think I understand now a bit better on the test failures being related to the tests themselves not necessarily what's going on within the upgrade (though still can be related in some cases).
I will try to rebase to the test channel and create a PR, even if it will never be merged. The Django LTS approach makes sense to me, too, even if the developer team count was slightly higher :).
I'll post here, once I have figured out the test channel rebasing (since I have never done that outside of the main branch I assume it's still fairly straightforward).

Fundamental question w.r.t. to existing ForeignKey relationships. I assume (in future Django versions) the on_delete attribute for ForeignKey should really be the "cascade" action, e.g., if a disk pool is deleted, things like shares, samba exports, etc. should also disappear from the database, right? Or are there scenarios where dependent entries should stick around (like in Rockons we keep some information after uninstall, unless one runs the cleanup script explicitly)?

**Comment by phillxnet** [comment]
@Hooverdan96 Regarding:

> Fundamental question w.r.t. to existing ForeignKey relationships. I assume (in future Django versions) the on_delete attribute for ForeignKey should really be the "cascade" action, e.g., if a disk pool is deleted, things like shares, samba exports, etc. should also disappear from the database, right?

We are likely going to have to prove this on a case by case basis but for now I say we go with the default and see what breaks in the testing channel. The newer Django versions have just made this setting explicit is my understanding so if we continue as-is and simply explicitly state the default we can take it from there. Unless of course we already do state our behaviour as being different from default but I don't remember seeing this anywhere.

Nice trick regarding the pg_dump cheers.

**Comment by phillxnet** [comment]
So the latest on this so far is that I've tried creating a new fixture with the expected pool 'in-situ' but as was the case when we had to mock objects out to fix all the unit tests way back, the objects resulting from fixtures just don't seem to exist. And given the new Django is way more picky about object interconects, the by-hand mocking of objects is now unsustainable. I'll update on any progress and do please chip-in if you have any pointers as to why our fixtures are not resulting in objects. Specifically the ones we are trying to mock here (Pool objects for starters) should simply exist. I know of reloading objects from db to refresh our object state from fixture but this I think should be implicit in our testing setup. Plus: a long time ago, in a code base we are build on, many of these mocks were not even necessary; they were instantiated by the fixtures. So I think this issue should get to the root cause of why our previously manageable mocks are:
1. require when they exist in a fixture (or did when I made a fresh one).
2. how we can replace all that are required by fixtures.
I'll submit my dabblings to date as I have to tent to some other issues before continuing with this branch anyway.

**Comment by phillxnet** [comment]
The commit associated with this issue to date contains in-code instructions for some of the associated fixtures we are considering here. I propose that we add such instructions to all tests that depend on fixtures so that as and when we need to re-do or even re-define fixtures we have the instructions current in the associated unit test file.

My current suspicion is that we have a test settings configuration somewhere that is overriding our fixture definition in each unit test. Seems unlikely but these tests used to resource their fixtures and did work way back with those fixtures. Plus we need to refresh all fixtures. But in the case of the currently examined tests the fixture had already been regenerated relatively recently.

Main fixtures that need updating would be those associate with our pre by-id device name days. Plus some of our existing fixtures have way too much junk in. We need them to be as small as possible for the given task. They can then more easily be hand edited if that is all that is required at any time in the future.

**Comment by phillxnet** [comment]
@FroggyFlox Just to note this is where, for the time being, I've popped in my fix for the django settings so we can run such things as, in this case, the db dumps (fixture creation process).

**Comment by FroggyFlox** [comment]
Thanks a lot for the commit, @phillxnet, especially with the included notes and instructions; at least it'll make sure there's no confusion and divergence on my end on how I was preparing the fixtures.

I remember banging my head on the table with these fixture issues a while ago but your pointers here re-motivated me to have another look/attempt.

Thanks again!

**Comment by phillxnet** [comment]
This issue has been mostly superseded by #2384 given #2382 is now resolved.

It is likely that #2384's resolution will also resolve this issue given it is moving our tests, whole-sale, over to a fixture based approach.

**Comment by phillxnet** [comment]
As suspected within the exploration within the comments of this issue. The mass test failure was down to a new sensitivity in our recently upgraded Django. And a move to our recently re-enabled fixture capability (Thanks again to @FroggyFlox) and we have
this issue as being:
Fixed by #2400
